### PR TITLE
Hotfix/output valid memoryaddress

### LINKED ISSUE #194: Output components can cause crashes
There is no guard in the output components to stop them from accessing invalid memory adresses.

### PR DESCRIPTION
**Pull request by riedersa**
Closes #194 
Now only valid Memory addresses schould be possible.

### CODE REVIEW

File: source/ui/Components/Output/ConsoleSettingsWindow.qml [hunk 1]
```diff
@@ -91,9 +91,12 @@ Window {
 
                 // Reads the current input and passes the new value to the model.
                 function processInput() {
-                    var inputValue = controlsColumn.integerFromInputString(String(baseAddressTextField.text))
-                    if (inputValue !== undefined && inputValue >= 0) {
+                    var inputValue = controlsColumn.integerFromInputString(String(baseAddressTextField.text));
+                    var maxSize = outputComponent.getMemorySize();
+                    if (inputValue !== undefined && inputValue >= 0 && inputValue <= maxSize) {
                         outputComponent.setOutputItemProperty(outputItemIndex, "baseAddress", inputValue);
+                    }else{
```
Comment (goldsborough) [inline source/ui/Components/Output/ConsoleSettingsWindow.qml]: `} else {` (spaces)

File: source/ui/Components/Output/ConsoleSettingsWindow.qml [hunk 2]
```diff
@@ -128,7 +131,7 @@ Window {
 
                 // Reads the current input and passes the new value to the model.
                 function processInput() {
-                    var inputValue = !ab.checked;
+                    var inputValue = ab.checked? 0 : 1;
```
Comment (goldsborough) [inline source/ui/Components/Output/ConsoleSettingsWindow.qml]: one space before `?`

File: source/ui/Components/Output/LightstripSettingsWindow.qml [hunk 3]
```diff
@@ -85,8 +85,17 @@ Window {
...
+                    var maxSize = outputComponent.getMemorySize();
+                    if (inputValue !== undefined && inputValue >= 0 && inputValue <= maxSize) {
                         outputComponent.setOutputItemProperty(outputItemIndex, "baseAddress", inputValue);
+                        var maxStrips = (outputComponent.getMemorySize() - (inputValue)) * 8;
+                        var strips = controlsColumn.integerFromInputString(String(numberOfStripsTextField.text));
+                        if(strips > maxStrips){
+                            numberOfStripsTextField.text = maxStrips + "";
+                            numberOfStripsTextField.processInput();
+                        }
+                    }else{
```
Comment (goldsborough) [inline source/ui/Components/Output/LightstripSettingsWindow.qml]: `} else {`

File: source/ui/Components/Output/LightstripSettingsWindow.qml [hunk 4]
```diff
@@ -106,9 +115,14 @@ Window {
                 // Reads the current input and passes the new value to the model.
                 function processInput() {
                     var inputValue = controlsColumn.integerFromInputString(String(numberOfStripsTextField.text));
-                    if (inputValue !== undefined && inputValue > 0) {
+                    var size = outputComponent.getMemorySize();
+                    var maxStrips = (outputComponent.getMemorySize() - size) * 8;
+                    if (inputValue !== undefined && inputValue > 0 && inputValue <= maxStrips) {
                         outputComponent.setOutputItemProperty(outputItemIndex, "numberOfStrips", inputValue);
                     }
+                    else{
```
Comment (goldsborough) [inline source/ui/Components/Output/LightstripSettingsWindow.qml]: `else {`

File: source/ui/output-component.cpp [hunk 5]
```diff
@@ -123,3 +123,8 @@ OutputComponent::getTextFromMemory(int start, QString currentText, int mode) {
   }
   return QString::fromStdString(text);
 }
+
+int OutputComponent::getMemorySize() {
+  int size = _memoryAccess.getMemorySize().get();
```
Comment (goldsborough) [inline source/ui/output-component.cpp]: just `return _memoryAccess.getMemorySize().get();`

**Review by goldsborough — APPROVED** [review]

File: source/ui/Components/Output/LightstripSettingsWindow.qml [hunk 6]
```diff
@@ -106,8 +115,12 @@ Window {
                 // Reads the current input and passes the new value to the model.
                 function processInput() {
                     var inputValue = controlsColumn.integerFromInputString(String(numberOfStripsTextField.text));
-                    if (inputValue !== undefined && inputValue > 0) {
+                    var size = outputComponent.getMemorySize();
+                    var maxStrips = (outputComponent.getMemorySize() - size) * 8;
```
Comment (FelixOpolka) [inline source/ui/Components/Output/LightstripSettingsWindow.qml:119]: `maxStrips` will always be 0, as you calculate `(outputComponent.getMemorySize() - outputComponent.getMemorySize()) * 8`. Guess it should be `size - controlsColumn.integerFromInputString(String(baseAddressTextField.text))`.

File: source/ui/Components/Output/LightstripSettingsWindow.qml [hunk 7]
```diff
@@ -106,8 +115,12 @@ Window {
...
+                    if (inputValue !== undefined && inputValue > 0 && inputValue <= maxStrips) {
```
Comment (FelixOpolka) [inline source/ui/Components/Output/LightstripSettingsWindow.qml]: I would suggest to check for too many digits separately and default to the highest possible value (maxStrips) in this case. I think that's more intuitive for the user.

File: source/ui/Components/Output/LightstripSettingsWindow.qml [hunk 8]
```diff
@@ -106,8 +115,12 @@ Window {
...
                         outputComponent.setOutputItemProperty(outputItemIndex, "numberOfStrips", inputValue);
+                    } else {
+                        updateSettings();
```
Comment (FelixOpolka) [inline source/ui/Components/Output/LightstripSettingsWindow.qml:127]: Good idea!

File: source/ui/Components/Output/LightstripSettingsWindow.qml [hunk 9]
```diff
@@ -85,8 +85,17 @@ Window {
                 // Reads the current input and passes the new value to the model.
                 function processInput() {
                     var inputValue = controlsColumn.integerFromInputString(String(baseAddressTextField.text))
-                    if (inputValue !== undefined && inputValue > 0) {
+                    var maxSize = outputComponent.getMemorySize();
+                    if (inputValue !== undefined && inputValue >= 0 && inputValue <= maxSize) {
```
Comment (FelixOpolka) [inline source/ui/Components/Output/LightstripSettingsWindow.qml]: I think that should be `inputValue < maxSize`. Entering 1024 currently leads to crash on my machine.

File: source/ui/Components/Output/ConsoleSettingsWindow.qml [hunk 10]
```diff
@@ -91,9 +91,12 @@ Window {
 
                 // Reads the current input and passes the new value to the model.
                 function processInput() {
-                    var inputValue = controlsColumn.integerFromInputString(String(baseAddressTextField.text))
-                    if (inputValue !== undefined && inputValue >= 0) {
+                    var inputValue = controlsColumn.integerFromInputString(String(baseAddressTextField.text));
+                    var maxSize = outputComponent.getMemorySize();
+                    if (inputValue !== undefined && inputValue >= 0 && inputValue <= maxSize) {
```
Comment (FelixOpolka) [inline source/ui/Components/Output/ConsoleSettingsWindow.qml]: Should be `inputValue < maxSize`. (Same as below)

**Review by FelixOpolka — APPROVED** [review]

**Pull request merged**